### PR TITLE
Add a nvrtc configuration for libcu++

### LINKED ISSUE #141: [FEA]: Add NVRTC testing to libcudacxx
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

libcu++

### Is your feature request related to a problem? Please describe.

libcu++ is lacking coverage for NVRTC in the current matrix.

### Describe the solution you'd like

Add an extra job that launches tests for NVRTC in libcudacxx. Prioritize CUDA 12.X images, with 11.X images being extra credit.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by miscco**
We also want to test with nvrtc. 

The main problem is that we only ever want to test with a single configuration and neither cub and thrust.

For that I went with a new build job that should exclude all unwanted configurations

Fixes #141

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -44,36 +44,59 @@ jobs:
...
+          - cuda_version: "11.1"
+          - cuda_version: "12.1"
+          - compiler: clang
+          - compiler: gcc
+        include:
+          - cuda_version: "12.1"
+          - compiler: gcc-12
+    with:
+      per_cuda_compiler_matrix: ${{ toJSON(fromJSON(needs.compute-matrix.outputs.PER_CUDA_COMPILER_MATRIX)[ format('{0}-{1}', matrix.cuda_version, matrix.compiler) ]) }}
+      build_script: "./ci/build_libcudacxx_nvrtc.sh"
```
Comment (jrhemstad) [inline .github/workflows/pr.yml:88]: I thought building libcu++ with NVRTC and the NoopExecutor means that nothing happens at all?

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -44,36 +44,59 @@ jobs:
...
+      test_script: "./ci/test_libcudacxx_nvrtc.sh"
```
Comment (jrhemstad) [inline .github/workflows/pr.yml:89]: I'm thinking that we don't really need separate build/test jobs for the NVRTC testing.

The only reason I did that for other environments is to preseed sccache with the build artifacts using CPU-only build machines to avoid paying the full compile cost on GPU machines (which are a scarcer resource). 

Given NVRTC doesn't work with sccache, there's no need to do that here.

File: .github/workflows/pr.yml [hunk 3]
```diff
@@ -44,36 +44,59 @@ jobs:
...
+        cuda_version: ${{ fromJSON(needs.compute-matrix.outputs.CUDA_VERSIONS) }}
+        compiler: ${{ fromJSON(needs.compute-matrix.outputs.COMPILERS) }}
+        exclude:
+          - cuda_version: "11.1"
+          - cuda_version: "12.1"
+          - compiler: clang
+          - compiler: gcc
+        include:
+          - cuda_version: "12.1"
+          - compiler: gcc-12
```
Comment (jrhemstad) [inline .github/workflows/pr.yml:85]: This is effectively the same as just hardcoding the environments to use. Instead, we want to query from the matrix file what the newest CUDA version is and use that. I think it will be easier to not use the `build-and-test.yaml` workflow.

File: libcudacxx/.upstream-tests/utils/libcudacxx/test/config.py [hunk 4]
```diff
@@ -423,7 +423,7 @@ def configure_no_execute(self):
     def configure_ccache(self):
         use_ccache_default = os.environ.get('CMAKE_CUDA_COMPILER_LAUNCHER') is not None
         use_ccache = self.get_lit_bool('use_ccache', use_ccache_default)
-        if use_ccache:
+        if use_ccache and not self.cxx.is_nvrtc:
```
Comment (jrhemstad) [inline libcudacxx/.upstream-tests/utils/libcudacxx/test/config.py:426]: Why is this needed? Won't sccache just do nothing when using nvrtc?

Comment (miscco) [reply libcudacxx/.upstream-tests/utils/libcudacxx/test/config.py:426]: sccache will hard error about an unsupported compiler

File: .github/workflows/pr.yml (hunk 3 above)
Comment (jrhemstad) [reply .github/workflows/pr.yml:85]: https://github.com/NVIDIA/cccl/issues/204

I'm going to do this first which will make this PR easier.

File: libcudacxx/.upstream-tests/utils/libcudacxx/test/config.py (hunk 4 above)
Comment (miscco) [reply libcudacxx/.upstream-tests/utils/libcudacxx/test/config.py:426]: We have a nvrtc.sh which does some dark magic

**Comment by miscco** [comment]
The iages are currently still running 12.0 drivers, so this PR is currently blocked until we get those images upgraded, presumably after the 23.08 release of rapids

File: libcudacxx/.upstream-tests/utils/libcudacxx/test/config.py (hunk 4 above)
Comment (griwes) [reply libcudacxx/.upstream-tests/utils/libcudacxx/test/config.py:426]: @jrhemstad exactly. We stitch together a source file that does that and includes the source of a given test, and we build and run that as the test executable with lit.

**Review by wmaxey — APPROVED** [review]

**Comment by wmaxey** [comment]
Can merge after conflict resolution, the changes look good.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
We'll want to hold off on merging this until I reconcile the new matrix format.

**Review by jrhemstad — APPROVED** [review]
This should be good to go now.

**Comment by jrhemstad** [comment]
> I assume we also want this for 2.2?

Since this is all just infrastructure and doesn't impact user code, I don't think it is important to get it into the 2.2 release.

**Review by jrhemstad — APPROVED** [review]
